### ISSUE TITLE
Find Empty namespaces

### ISSUE THREAD

**Issue by yonahd** [yonahd/kor#92]
Find namespaces with no resources

**Comment by liornoy** [comment]
hey @yonahd I would like to work on that, please.
Is it fair to say that we call a namespace empty if there are not resources in there at all? is that correct?

**Comment by yonahd** [comment]
Yes. It should be with no resources except for the default service account

**Comment by liornoy** [comment]
Thanks, I already started working on this, and on my Kind cluster there is an auto-generated "kube-root-ca.crt' configmap that we might also want to add as an exception. wdyt?

**Comment by liornoy** [comment]
I see we even have a `exceptionconfigmaps` var for it

**Comment by yonahd** [comment]
> Thanks, I already started working on this, and on my Kind cluster there is an auto-generated "kube-root-ca.crt' configmap that we might also want to add as an exception. wdyt?

Yes. This is default as well

**Comment by roopeshsn** [comment]
Hi @yonahd, I'd like to address this issue. Are the default resources, such as the service account and `kube-root-ca.crt` configmap, consistent across Kubernetes distributions? If so, can we determine whether a namespace is unused by checking for the presence of these default resources, and concluding that it is unused if there are no additional resources in that namespace?

**Comment by yonahd** [comment]
> Hi @yonahd, I'd like to address this issue. Are the default resources, such as the service account and `kube-root-ca.crt` configmap, consistent across Kubernetes distributions? If so, can we determine whether a namespace is unused by checking for the presence of these default resources, and concluding that it is unused if there are no additional resources in that namespace?

This is consistent. We want to take an approach similar to https://github.com/corneliusweig/ketall/tree/master

**Comment by roopeshsn** [comment]
I took time to view the implementation of Ketall. It constructs a slice of resource names with the API resources (`[configmaps endpoints pods ...]`). It creates a [new resource builder](https://github.com/corneliusweig/ketall/blob/16390bd75095091b59142cd34f4c5ae060738556/internal/client/client.go#L167) from the `kubernetes/cli-runtime` package and calls the methods `Do().Object()` which prints the resources from all the namespaces or from a specific namespace when the `--namespace=<namespace>` flag is passed.

Do you want me to see the logic of Do().Object()?

**Comment by roopeshsn** [comment]
The thing I am confused about is that though we find all the resource types, create a builder from `cli-runtime`, and pass the resource types it will return an object with all the namespaced and default resources in a namespace. How it will be useful to find an unused namespace?

**Comment by yonahd** [comment]
@roopeshsn We don't find all resources.
We don't even interact with all of them. 
The user can know which namespace he can consider empty(and avoid thinking its used by the default resources) and delete it. Not sure I completely understand the question

**Comment by roopeshsn** [comment]
> @roopeshsn We don't find all resources. We don't even interact with all of them. The user can know which namespace he can consider empty(and avoid thinking its used by the default resources) and delete it. Not sure I completely understand the question

Exactly! Ketall finds the namespace resources using the `cli-runtime` package. How does that implementation help to find an unused namespace? However, we need to check and conclude a namespace is empty if it contains only default resources.

**Comment by isindir** [comment]
The above PR is still WIP - lacks parallel namespace processing and unit tests

**Comment by isindir** [comment]
Added some unit tests and recently rabased once more, @yonahd would it be possible for you to review and either accept or refuse this PR ?
